### PR TITLE
feat(932150): apply same philosophy behind 932100

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

This PR is the start of getting a new 932150 rule.

It will behave similarly to the 932100 movements we did back at the dev retreat. Mainly:

- [x] 932150 -> 932250 + 932260. First for up to 3 chars, and then a generic rule in 932260.
- [x] 932151: removed rule and moved tests to 932250 + 932260
- [x] Split tests in 932250 for < 4 chars and >= 4 for 932260 d346307d2d915ec0daeb7b3014f927df25a25a72
- [x] Verify that we cover all remaining BB issues

### CODE REVIEW

**Comment by fzipi** [comment]
@theseion This is the start. We can do it.

**Comment by dune73** [comment]
What are the open questions around 932151? Does a decision depend on the detection-performance of 932250 + 932260?

**Comment by fzipi** [comment]
> What are the open questions around 932151? Does a decision depend on the detection-performance of 932250 + 932260?

I didn't even take a look at it. But I guess being a strict sibling based on different words might or might not make sense under the new splitting on length. And it could be merged part in 250 and part in 260.

**Comment by dune73** [comment]
I follow your reasoning on 931151. Let's hold on for the time being and see where the rest of this goes.

**Comment by dune73** [comment]
This sounds useful. In general and here especially.

**Comment by dune73** [comment]
@fzipi can you give a status update?

It would also be useful to link the PRs to the (checkboxed) tasks listed in the issue.

**Comment by fzipi** [comment]
> @fzipi can you give a status update?
> 
> It would also be useful to link the PRs to the (checkboxed) tasks listed in the issue.

What you are seeing is the status update. I call it "Update", but gladly can rename it. Maybe @theseion can tell us if he is working on that or what is happening now.

**Comment by dune73** [comment]
Thanks

**Comment by theseion** [comment]
I have a prototype implementation of a new `include-except` processor. Right now, I'm working on a bug that occurs when including files (high priority).

**Comment by theseion** [comment]
@fzipi 932150 contains some entries that don't exist in our unix-shell-data files:

```
bunzip2
bzcmp
bzexe
bzz
core_perl/zipdetails
GET
gzexe
lz
lz4
pigz
pxz
unlz4
unpigz
unzstd
zstd
zstdcat
zstdgrep
zstdless
zstdmt
```

Not sure wether we should add them.

**Comment by theseion** [comment]
Conversly, this is the list of words that we would currently exclude. I suggest we add these  to an exclude file and use the new `#include-except` processor.

```
ab
apt
ar
arj
arp
as
ash
at
awk
aws
c89
cat
cmp
cp
cut
dd
dig
dnf
du
eb
ed
eqn
ex
fc
fi
fmt
ftp
gdb
gem
ghc
git
go
grc
hd
id
ip
jjs
jq
ld
ldd
ln
lp
lua
man
mtr
mv
net
nl
nm
npm
od
pax
pdb
pf
pg
pic
pip
pkg
pr
pry
ps
ptx
rar
red
rev
rm
rpm
set
sg
ss
su
tac
tbl
tee
tex
tic
top
ul
vi
vim
wc
who
xxd
yum
adduser
agetty
alias
alpine
ansible-playbook
apt-get
arch
aria2c
ascii-xfr
ascii85
aspell
atobm
batch
bpftrace
breaksw
bridge
bundler
busctl
byebug
cancel
capsh
certbot
chattr
chdir
check_by_ssh
check_cups
check_log
check_memory
check_raid
check_ssl_cert
check_statusfile
chflags
chmod
choom
chown
chroot
cobc
column
comm
composer
compress
cowsay
cowthink
cpan
cpio
cpulimit
crash
crontab
csplit
csvtool
cupsfilter
date
dhclient
dialog
dmidecode
dmsetup
docker
done
dosbox
dpkg
dvips
easy_install
efax
emacs
endif
endsw
env-update
esac
exiftool
expand
expect
export
expr
facter
file
find
finger
fish
flock
fold
foreach
fping
function
gawk
gcore
genie
genisoimage
getfacl
ghci
gimp
ginsh
gtester
hexdump
highlight
history
hostid
hostname
hping3
htdigest
htpasswd
iconv
ifconfig
iftop
install
ionice
ip6tables
ipconfig
iptables
ispell
jexec
jobs
join
journalctl
jrunscript
kill
killall
knife
ksshell
last
latex
ldconfig
less
links
local
locate
loginctl
logname
logsave
look
lshw
ltrace
lualatex
luatex
lwp-dump
lwp-mirror
lwp-request
mail
mailx
make
mawk
mkdir
more
mosquitto
mount
msgattrib
msgcat
msgconv
msgfilter
msgmerge
msguniq
mutt
mysql
nano
nasm
nawk
neofetch
nice
nmap
node
nroff
nsenter
nslookup
octave
openssl
openvpn
openvt
opkg
passwd
paste
patch
pdflatex
pdftex
perf
perlsh
perms
pico
pidstat
ping
pkg_info
pkginfo
psftp
psql
python
puppet
pushd
rake
raku
readelf
redcarpet
restic
rlogin
rlwrap
route
rpmdb
rpmquery
rpmverify
run-mailcap
run-parts
rview
rvim
sash
screen
script
service
setarch
setfacl
shuf
shutdown
sleep
slsh
smbclient
snap
soelim
sort
split
sqlite3
ssh-keygen
ssh-keyscan
sshpass
start-stop-daemon
stdbuf
strace
systemctl
systemd-resolve
task
taskset
tclsh
tcpdump
tftp
time
timedatectl
tmux
touch
troff
tshark
ulimit
unexpand
uniq
unlink
unshare
update-alternatives
uudecode
uuencode
valgrind
view
vimdiff
virsh
volatility
wall
watch
whiptail
whois
wireshark
wish
xelatex
xetex
xmodmap
xmore
xpad
xterm
yarn
yelp
zathura
zsoelim
zypper
```

**Comment by dune73** [comment]
@fzipi, this is in your hands, is not it? Is there anything I can do?

**Comment by theseion** [comment]
I'll finish this tomorrow (hopefully).

**Comment by dune73** [comment]
Nice.

Stupid question: "Finishing" includes "Verify that we cover all remaining BB issues", correct?

**Comment by theseion** [comment]
That's the plan, yes.

**Comment by theseion** [comment]
@fzipi can you review?

**Review by fzipi — COMMENTED** [review]
Awesome job Max. Finally.

**Comment by dune73** [comment]
Wow, that's a lot of stuff. Thank you very much.

I see that all the tests are passing; a very good sign apparently, also because this PR is too big to really, really review word for word, so I trust our tests in terms of "do the rules do the things they have to".

Still, I looked through it in detail, and took a few notes / questions:

* You are naming 932101 and 932106 as strict siblings in your PR. They are meant to move to the 9322xx range as well, are not they?
* You had to touch on 932230 / 40 too. Why?
* I have not yet understood the solution to the FP problem. I see how we have a list of all unix commands split in <4 and >=4 char lengths, and then a single list of FP unix command that are excluded from the inclusion. However are we protecting against payloads targetting these Unix commands at all, or do we rely on secondary indicators like the ones in 932240 for them?

This is the only stuff that occurred to me. Again, my insight is limited, but I like the general architecture and how you put it into practice here.

**Comment by theseion** [comment]
> You are naming 932101 and 932106 as strict siblings in your PR. They are meant to move to the 9322xx range as well, are not they?

Sure, I can do that.

> You had to touch on 932230 / 40 too. Why?

I've added process substitution detection, e.g., `<()`. That required a change to the evasion regex used by the `cmdline` processor, which in turn impacts all rules that use `cmdline unix`.

> I have not yet understood the solution to the FP problem. I see how we have a list of all unix commands split in <4 and >=4 char lengths, and then a single list of FP unix command that are excluded from the inclusion. However are we protecting against payloads targetting these Unix commands at all, or do we rely on secondary indicators like the ones in 932240 for them?

If I understand you question, you are asking about the exclusion list. 932230 and 932235 use the full command words list without exclusions. There, we detect all of them. 932250 and 932260 use exclusions to reduce FP's. At PL2 we might still detect some of the excluded ones if they are matched by 932240.
In #3124 we could remove the exclusions again from the proposed sibling rule at PL2, in which case we would match the full command list too. Judging from the comments on Slack I think we might do that. However, it would probably make sense to then merge that rule with 932236, which does the same thing already but with the prefix from 932230. So, instead of having two rules that match the same thing, but once with, once without prefix, we'd just have one rule without prefix matching.

In conclusion: I propose to drop #3124 and instead adjust 932236 as follows:
- drop the prefix match
- match all targets from 932230 in addition to `Referer` and `User-Agent` headers

Does that make sense to you?

**Comment by dune73** [comment]
Thank you for these explanations.

* +1 on the moving of rule IDs 932101 and 932106
* Thanks for explanations on 932230 / 40. Makes perfect sense.
* And finally thanks for the outlining of the various rules next to one another. I'm getting a better overview now.  Personally I agree on the idea to drop #3124 and to adjust 932236 instead.

Happy to approve this PR once this is implemented.

@fzipi would you mind adding your comments here as well? Are you OK with the PR?

**Comment by fzipi** [comment]
Yes, I like the idea of unifying and dropping #3124.

**Comment by fzipi** [comment]
I cannot approve explicitly because I've started this PR. But here is my written approval.

**Comment by theseion** [comment]
I've updated the ID's for 932101 and 932106 to 932231 and 932232. I hope those are good.

Due to the complexity of these changes I don't want to add any more changes to this PR. The other changes we've discussed will follow in separate PR's.

@dune73 Can you please approve the PR? Otherwise, I can't merge.

**Review by dune73 — APPROVED** [review]

**Comment by dune73** [comment]
Thank you guys. Let's merge.
